main/src/com/google/refine/io/FileProjectManager.java

```diff
@@ -288,15 +290,20 @@ protected void saveWorkspace() {
                 file.renameTo(oldFile);
             }
 
+            if (oldFile.exists() && _projects.isEmpty()) {
+                oldFile.delete();
+            }
```

Comment:
It is not clear to me why we need to delete the old file in this case, can you explain a bit more?

Reply:
When a workspace is saved after removing a projectID its previous state is stored in wospace.old.json. When the only existing project is removed and the workspace is saved, workspace.old.json while retain the deleted projectID. 
On restarting the server, an empty project directory for the projectID  in workspace.old.json is created and this brings back the bug we are fixing.

Reply:
But why is the empty project directory leftover? Should we not rather delete that directory instead of deleting `workspace.old.json`?

Reply:
The project directory is actually deleted as expected but when the server is restarted projects are loaded from :https://github.com/OpenRefine/OpenRefine/blob/1cb874819284ae21a43117decc5f10297a6092a7/main/src/com/google/refine/io/FileProjectManager.java#L361-L372

Since `workspace.old.json` has a projectID and there is no existing project directory for it an empty project directory is created

Reply:
Ah, then that sounds like the `loadFromFile` method incorrectly considers that loading `workspace.json` failed if it did not load any project: it should instead recognize that it successfully loaded an empty workspace, no?

Reply:
Correct. That should be the case

Reply:
Hi @WaltonG, is this still needed after your changes then?

Reply:
Hi @wetneb I don't think it's needed. Let me confirm

Reply:
Hi @wetneb, I have confirmed that loading of projects from workspace.temp.json and workspace.old.json is not required

Reply:
Hmm, that is curious! Intuitively, this fallback mechanism should still be useful if `workspace.json` disappeared (after a failed save, perhaps).
I would expect that this code to load from `workspace.temp.json` and `workspace.old.json` stays, but the return value of `loadWorkspace` is not whether a project was found or not, but rather whether the loading of the workspace succeeded or not (which is different, since it is possible to successfully load an empty workspace). This is what I have tried to explain in my comment above - is this new comment any clearer?

Reply:
Sure the comment is clearer and a fallback mechanism is actually needed. With that in mind am rechecking if deleting `workspace.old.json`  is still needed.

Reply:
> Hmm, that is curious! Intuitively, this fallback mechanism should still be useful if `workspace.json` disappeared (after a failed save, perhaps).

Considering this scenario, when a project directory is deleted but saving the workspace  failed then the fallback mechanism will reference a non existing project directory. In the following code is it ok to move `saveWorkspace()` to the synchronized block

https://github.com/OpenRefine/OpenRefine/blob/f151be913b5d7de25b312a48300080190e68a605/main/src/com/google/refine/io/FileProjectManager.java#L327-L338